### PR TITLE
Adding option to keep range data for missed points instead of zero to consistent interpolation op

### PR DESCRIPTION
**Pull request by sslattery**
This option lets users keep the part of the range vector that has
missing points when enabled. This automatically adds tracking of
missed range entities as well as this is needed for making the vector.
To enable this in a general way the map operator now always pulls
data from the range vector application before applying the implementation
operator. We will want to review the performance of this later.

### CODE REVIEW

**Comment by sslattery** [comment]
Damien can you review?

File: packages/Operators/test/tstConsistentInterpolationOperator.cpp [hunk 1]
```diff
@@ -302,7 +302,7 @@ TEUCHOS_UNIT_TEST( ConsistentInterpolationOperator, no_domain_0_test )
     for ( int i = 0; i < num_points; ++i )
     {
 	point_ids[i] = num_points*comm_rank + i;
-	point_dofs[i] = 0.0;
+	point_dofs[i] = 2.2;
```
Comment (dalg24) [inline packages/Operators/test/tstConsistentInterpolationOperator.cpp]: why did you change that one?

File: packages/Operators/test/tstConsistentInterpolationOperator.cpp [hunk 2]
```diff
@@ -868,5 +868,111 @@ TEUCHOS_UNIT_TEST( ConsistentInterpolationOperator, local_missed_range_test )
...
+
+    // Get the communicator.
+    Teuchos::RCP<const Teuchos::Comm<int> > comm =
+	Teuchos::DefaultComm<int>::getComm();
+    int comm_size = comm->getSize();
+    int comm_rank = comm->getRank();
+
+    // DOMAIN SETUP
+    // Don't put domain entities on proc 0.
+    int num_boxes = (comm->getRank() != 0) ? 5 : 0;
```
Comment (dalg24) [inline packages/Operators/test/tstConsistentInterpolationOperator.cpp]: use ``comm_rank`` here

File: packages/Operators/test/tstConsistentInterpolationOperator.cpp (hunk 1 above)
Comment (sslattery) [inline packages/Operators/test/tstConsistentInterpolationOperator.cpp]: It was to make sure we would overwrite the data with zeros if we missed points. The test wasn't checking that before.

File: packages/Operators/test/tstConsistentInterpolationOperator.cpp [hunk 3]
```diff
@@ -868,5 +868,111 @@ TEUCHOS_UNIT_TEST( ConsistentInterpolationOperator, local_missed_range_test )
...
+    Teuchos::Array<DataTransferKit::SupportId> box_ids( num_boxes );
+    Teuchos::ArrayRCP<double> box_dofs( num_boxes );
+    Teuchos::Array<Entity> boxes( num_boxes );
+    for ( int i = 0; i < num_boxes; ++i )
+    {
+	box_ids[i] = num_boxes*(comm_size-comm_rank-1) + i;
+	box_dofs[i] = 2.0*box_ids[i];
+	boxes[i] = BoxGeometry(box_ids[i],comm_rank,box_ids[i],
+		       0.0,0.0,box_ids[i],1.0,1.0,box_ids[i]+1.0);
```
Comment (dalg24) [inline packages/Operators/test/tstConsistentInterpolationOperator.cpp]: So basically boxes 1x1x1 stacked on top of each other in z-direction, right?

File: packages/Operators/test/tstConsistentInterpolationOperator.cpp [hunk 4]
```diff
@@ -868,5 +868,111 @@ TEUCHOS_UNIT_TEST( ConsistentInterpolationOperator, local_missed_range_test )
...
+    Teuchos::Array<Entity> points( num_points );
+    for ( int i = 0; i < num_points; ++i )
+    {
+	point_ids[i] = num_points*comm_rank + i;
+	point_dofs[i] = 2.2;
+	point[0] = 0.5;
+	point[1] = 0.5;
+	point[2] = point_ids[i] + 0.5;
+	points[i] = Point(point_ids[i],comm_rank,point);
+    }
```
Comment (dalg24) [inline packages/Operators/test/tstConsistentInterpolationOperator.cpp]: 5 points per proc, located in the center of boxes of proc+1, except the last proc that will be missed

File: packages/Operators/test/tstConsistentInterpolationOperator.cpp [hunk 5]
```diff
@@ -868,5 +868,111 @@ TEUCHOS_UNIT_TEST( ConsistentInterpolationOperator, local_missed_range_test )
...
+    map_op->setup( 
+	domain_manager.functionSpace(), range_manager.functionSpace() );
+
+    // Apply the map.
+    map_op->apply( *domain_dofs, *range_dofs );
+
+    // Check the results of the mapping.
+    for ( int i = 0; i < num_points; ++i )
+    {
+	double test_val = (comm_rank != comm_size-1) ? 2.0*point_ids[i] : 2.2;
```
Comment (dalg24) [inline packages/Operators/test/tstConsistentInterpolationOperator.cpp]: ``point_ids[i]`` just happen to be equal to the id of the box it was found in.
That does the work but man it's not easy to read...

File: packages/Operators/test/tstConsistentInterpolationOperator.cpp [hunk 6]
```diff
@@ -868,5 +868,111 @@ TEUCHOS_UNIT_TEST( ConsistentInterpolationOperator, local_missed_range_test )
...
+	TEST_EQUALITY( test_val, point_dofs[i] );
+    }
+
+    // Check that proc zero had all points not found.
```
Comment (dalg24) [inline packages/Operators/test/tstConsistentInterpolationOperator.cpp]: no actually the proc with the highest rank is the one that has the points out of the boxes

File: packages/Operators/test/tstConsistentInterpolationOperator.cpp [hunk 7]
```diff
@@ -868,5 +868,111 @@ TEUCHOS_UNIT_TEST( ConsistentInterpolationOperator, local_missed_range_test )
...
+    int num_missed = (comm_rank != comm_size-1) ? 0 : 5;
```
Comment (dalg24) [inline packages/Operators/test/tstConsistentInterpolationOperator.cpp]: seriously, a negation in a ternary operator?

File: packages/Operators/test/tstConsistentInterpolationOperator.cpp [hunk 8]
```diff
@@ -868,5 +868,111 @@ TEUCHOS_UNIT_TEST( ConsistentInterpolationOperator, local_missed_range_test )
...
+    Teuchos::Array<EntityId> missed_ids(
+	map_op->getMissedRangeEntityIds() );
+    TEST_EQUALITY( missed_ids.size(), num_missed );
+    std::sort( point_ids.begin(), point_ids.end() );
+    std::sort( missed_ids.begin(), missed_ids.end() );
+    for ( int i = 0; i < num_missed; ++i )
+    {
+	TEST_EQUALITY( missed_ids[i], point_ids[i] );
```
Comment (dalg24) [inline packages/Operators/test/tstConsistentInterpolationOperator.cpp]: don't you want to check the dofs/values rather than the ids?

**Comment by dalg24** [comment]
Here are a few other comments:
* We need to document more clearly what's going on here: If ``"Keep Missed Range Data"`` is present in the list of parameters we will leave the corresponding dofs/values UNCHANGED in the target vector.
* Do we want to allow for a user specified value to be set in the missed entities? Or is the user responsible for setting the target vector to that value before he calls apply? In that case we can add a comment to explain how that can be done.

**Comment by dalg24** [comment]
NOTE: I referenced the PR in the other issue by mistake. I removed the ref but it is still showing up on the PR. I'm not sure how to remove that.

File: packages/Operators/test/tstConsistentInterpolationOperator.cpp (hunk 1 above)
Comment (dalg24) [inline packages/Operators/test/tstConsistentInterpolationOperator.cpp]: ok

File: packages/Operators/test/tstConsistentInterpolationOperator.cpp (hunk 8 above)
Comment (sslattery) [inline packages/Operators/test/tstConsistentInterpolationOperator.cpp]: Here we are making sure that the missing ids match the point ids - not the values.

Comment (dalg24) [inline packages/Operators/test/tstConsistentInterpolationOperator.cpp]: OK you check values just before
